numpy/_core/src/multiarray/ctors.c

```diff
@@ -2405,6 +2409,63 @@ PyArray_FromInterface(PyObject *origin)
 }
 
 
+
+/*
+ * Returns -1 and an error set or 0 with the original error cleared, must
+ * be called with an error set.
+ */
+static inline int
+check_or_clear_and_warn_error_if_due_to_copy_kwarg(PyObject *kwnames)
+{
+    if (kwnames == NULL) {
+        return -1;  /* didn't pass kwnames, can't possibly be the reason */
+    }
+    if (!PyErr_ExceptionMatches(PyExc_TypeError)) {
+        return -1;
+    }
+
+    PyObject *type, *value, *traceback;
+    PyErr_Fetch(&type, &value, &traceback);
+    if (value == NULL) {
+        PyErr_Restore(type, value, traceback);
```

Comment:
I think a `goto restore` here and below might be easier.

Reply:
true, one of the advantages of having the helper now that this is trivial :).  (I had much fewer restores in a first iteration...)